process: add intial change management process (#443)
eclipse-score/score: docs/platform_management_plan/change_management.rst

   # SPDX-License-Identifier: Apache-2.0
   # *******************************************************************************

.. document:: Change Management Plan
   :id: doc__platform_change_management_plan
   :status: draft
   :safety: ASIL_B
   :tags: platform_management

> Change management / Change Management Plan

Reviewer: ```suggestion
Change Management / Change Management Plan
```

Author: updated as suggested

* Implementation of Change Requests is confirmed.
* Change Requests are tracked to closure and status of Change Requests is communicated to
  affected parties.
* Requests for Changes are properly documented.

Approach
++++++++

Change Request Execution
^^^^^^^^^^^^^^^^^^^^^^^^

Contributions in general to the **S-CORE** project are described here
(compare :need:`gd_guidl__contr_request_guideline`).

A Change Request is a specific contribution, and
it is the **ONLY** way to contribute new features or to modify the scope of existing features
in the **S-CORE** project.

Change Request Infrastructure and Types
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

:need:`GitHub Issues (ISSUE) <gd_guidl__issue_guideline>` are used for managing Change Requests.
The tool is used to create, plan, control, and monitor Change Requests within **S-CORE**.

:need:`GitHub Pull Requests (PR) <gd_guidl__pull_request_guideline>` are used for the implementation
of Change Requests. The tool is used to implement and verify Change Requests within **S-CORE**.

The next figure gives an overview, how Change Requests are realized in **S-CORE**. An ISSUE is

> used to create a Change Request including reuqired attributes as defined in

Author: coreccted

:ref:`chm_process_change_request_attributes`.
The ISSUE may be linked to SUB-ISSUEs, if required, to manage more complex Change Requests.

> Implementation of Change Requests requires at least one or more PRs linked to the ISSUE created

Reviewer: "At least one" already implicates that it can be more.

```suggestion
The implementation of a Change Requests requires at least one PR linked to the ISSUE created
```

for the Change Request.

.. figure:: _assets/score_change_request_overview.drawio.svg
  :width: 100%
  :align: center
  :alt: Change Request Overview

  Change Request Overview

Changes are clustered in the following types:

.. list-table:: Change Request Types
   :header-rows: 1
   :widths: 15,85,15

   * - Type
     - Description
     - Infrastructure
   * - Feature
     - Created by :need:`Contributor <rl__contributor>` to change requirements and work products, new feature
     - ISSUE with label ``feature_request``
   * - Modification
     - Created by :need:`Contributor <rl__contributor>` to change requirements and work products, scope change
     - ISSUE with label ``feature_modification``

> Change Request Attributes

Author: here are the labels defined, which are not part of the process part, so concrete realisation in the used tool, versus just type definition

^^^^^^^^^^^^^^^^^^^^^^^^^
:ref:`chm_process_change_request_attributes` are implemented as follows:

:need:`[[title]] <gd_req__change__attr_uid>` is identical to the ISSUE number.

:need:`[[title]] <gd_req__change__attr_status>` is defined by the combination of the ISSUE state
and the states of the linked PRs.

:need:`[[title]] <gd_req__change__attr_title>` is identical to the ISSUE title.

:need:`[[title]] <gd_req__change__attr_impact_description>` is identical to the description filled
of an ISSUE.

:need:`[[title]] <gd_req__change__attr_impact_safety>` is part of the description.

:need:`[[title]] <gd_req__change__attr_impact_security>` is part of the description.

:need:`[[title]] <gd_req__change__types>` is defined by labe of a ISSUE.

:need:`[[title]] <gd_req__change__attr_affected_wp>` is defined in the linked PR or part of the
description.

:need:`[[title]] <gd_req__change__attr_milestone>` is defined by the Milestone of a ISSUE.

Change Request Workflow
^^^^^^^^^^^^^^^^^^^^^^^

In General, every Change Request follows the following steps:

* 1. Create the Change Request
* 2. Review the Change Request
* 3. Approve the Change Request

> To 1. Create the Change Request:

Author: done, also for 2. and 3.

An ISSUE is the **ONLY** way to create and manage a Change Request in **S-CORE**.
A PR is the **ONLY** way to implement a Change Request in **S-CORE**,
thus an ISSUE must be linked at least to one or more PRs.

The figure below shows the workflow for the simple case of a Change Request.

An ISSUE with the label according to the Change Request type is created in status ``OPEN``.
The title of the ISSUE reflects the potential change.
The Description of the ISSUE describes the details and potential impact. For the new feature

> request use the template Feature request: :ref:`chm_templates`.

Author: added

> request use the template Feature request: :ref:`chm_templates`.

Author: finally the work product feature request

> request use the template Feature request: :ref:`chm_templates`.

Reviewer 2: Still unclear - we have two templates, one in the CR process for the issue and one for the work product?

The ISSUE is linked to one PR in status ``DRAFT``, which may contain the
required affected work products based on defined templates or implementation and verification
proposals.

Planning is done by setting the milestone of the ISSUE accordingly.

As long as the :need:`Contributor <rl__contributor>` wants to modify the content of the Change
Request, the linked PR is kept in status ``DRAFT``.

Change request status: ``draft`` is implemented as
ISSUE status ``OPEN`` and PR status ``DRAFT``.

To trigger the next step the PR status is changed from ``DRAFT`` to ``OPEN``.

> .. figure:: _assets/score_change_request_workflow_simple.drawio.svg

Author: image updated with reject

  :width: 100%
  :align: center
  :alt: Change Request Simple Workflow Overview

  Change Request Simple Workflow Overview

To 2. Review the Change Request:

The Change Request is analyzed from the :need:`Contributor <rl__contributor>`

> together with the :need:`Committer <rl__committer>` and the results are documented in the ISSUE.

Author: change to PR

Unless the information is not sufficient, the related PR is kept in status ``OPEN``.

:ref:`chm_checklist` can help to verify whether the information is complete.

> The implementation part of the Change Request is reviewed according the checklists of the affected

Author: rephrased

work products. Verification of the implementation must be successful.
If the verification is not sufficient, the related PR is kept in status ``OPEN`` or may changed
back to ``DRAFT`` (compare :need:`gd_guidl__issue_guideline`).

Change request status: ``in review`` is implemented as
ISSUE status ``OPEN`` and PR status ``OPEN``.

To 3. Approve the Change Request:

> :need:`Committer <rl__committer>` decides finally, if the Change Request is accepted or rejected.

Author: changed, and made is consistent with the workflow, technical or module lead

:need:`Committer <rl__committer>` checks finally if the Change Request is completed and the
required verification measures are executed and sucessfully passed.

If ``approved``, the status of the concerned PRs change to ``MERGED``,
otherwise, if rejected, PR status changes to ``CLOSED``.

If ``approved`` the status of the ISSUE is ``CLOSED``.

> The figure below shows the workflow for a complex case of a Change Request.

Reviewer 2: I woulld actually say it is the normal case (if we only consider feature requests/modifications) because we need planning also in the module repos -> therefore we do not ony have sub-issues but also referenced issues in other repos! Could you include this?

Author: modified example 2 to reflect that

The ISSUE is linked to SUB-ISSUES, and the SUB-ISSUE is linked to a PR. In principle the Change
Request workflows applies for all SUB-ISSUEs independently. Finally the ISSUE must be closed
manually

.. figure:: _assets/score_change_request_workflow_complex_1.drawio.svg
  :width: 100%
  :align: center
  :alt: Change Request Complex Workflow Overview Case 1

  Change Request Complex Workflow Overview Case 1

> Another version of a more complex Change Request case is shwon the following figure.

Reviewer 2: typo shwon

Author: resolved

> Here the ISSUE is linked to multiple PRs. As the PRs may accepted not at the same time, e.g. default

Reviewer 2: I would not show this, I think it is clear from text already. And I would also say that our tooling should support not to switch the issue permanently from open to closed.

Author: removed

to priorization, the Change Request must be kept in the correct status by manually changing the
ISSUE status to ``OPEN`` unless the last linked PR is accepted.

.. figure:: _assets/score_change_request_workflow_complex_2.drawio.svg
  :width: 100%
  :align: center
  :alt: Change Request Complex Workflow Overview Case 2

  Change Request Complex Workflow Overview Case 2

Change Management SW Platform Work Products
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. list-table:: SW Platform work products
    :header-rows: 1

    * - work product Id
      - Link to process
      - Process status
      - Link to issue
      - Link to WP
      - WP status

>     * - :need:`wp__policies`

Reviewer 2: do we need this? what part of it touches the Change Management?

Author: removed
